records/2022-12-13.upgrade-database.md

```diff
@@ -0,0 +1,104 @@
+
+### 12
+
+There are [breaking changes](https://www.postgresql.org/docs/12/release-12.html#id-1.11.6.18.4) from 11 to 12. Unknown if any affect us.
+
+### 13.14
+
+There are also [breaking changes](https://www.postgresql.org/docs/13/release-13.html#id-1.11.6.14.4) from 12 to 13. Unknown if any affect us.
+
+## Upgrade approaches
+
+From [Upgrade your Amazon RDS for PostgreSQL or Amazon Aurora PostgreSQL database, Part 1: Comparing upgrade approaches](https://aws.amazon.com/blogs/database/part-1-upgrade-your-amazon-rds-for-postgresql-database-comparing-upgrade-approaches/) there are a few approaches we can use.
+
+### Option A: In-place upgrade
+This is in some ways the easiest, in that we essentially tell AWS to upgrade the database and it does it. The downside is that the database will be unavailable for most of the upgrade time, and hence Notify will be down. Also, if something goes wrong we would have to restore the database from a backup (and hope that works).
+
+### Option B: Snapshot approach
+In this approach we would make a copy of the database and upgrade that copy, then switch the app to use the new database. The benefit is that if something doesn't work we can quickly switch back to the old database. This approach still requires downtime during the process though, since data added to the old database would not be in the new database.
+
+### Option C: AWS DMS approach
+We can use the AWS Database Migration Service. Here we create a new database with the new version, then turn on DMS. The old data will be copied into the new database. Also, any data added during this time will also be copied over. We then stop Notify, let any remaining data go to the new database, point Notify at the new database, and restart.
+This is a fairly straightforward approach (as these things go) and will have less downtime than other methods. Note that DMS does not work with all data types or database configurations, so it might not be an option.
+
+### Option D: Replication-based approach using pglogical
+This approach use the `pglogical` database extension along with database replication sets (A replication set is way to mark a collection of tables to be replicated.). In many ways this is very similar to Option C, however in this case once the new database is populated it stays up to date faster, and there is no "wait for last entries to transfer" downtime. Rather we just have to turn off Notify, switch to the new database, and turn Notify back on (and then turn off the extension). There are restriction on what database configurations are acceptible, however, and this is a more manual process than using DMS.
+
+### Discussion
+
+We want to minimize downtime and maximize probability of success. 
+- There probably isn't too much to be gained performance-wise by just moving to 11.16, so we should move to 12.12 or 13.14. Just moving to 12.12 will increase the probability of success, at the expense of having to do anothe migration at a future point.
```

Comment:
"at the expense of having to do anothe migration at a future point" - Missing 'r' in another